torchvision/transforms/v2/functional/_color.py

```diff
@@ -325,14 +325,21 @@ def _hsv_to_rgb(img: torch.Tensor) -> torch.Tensor:
     p = one_minus_s.mul_(v).clamp_(0.0, 1.0)
     i.remainder_(6)
 
-    mask = i.unsqueeze(dim=-3) == torch.arange(6, device=i.device).view(-1, 1, 1)
+    vpqt = torch.stack((v, p, q, t), dim=-3)
 
-    a1 = torch.stack((v, q, p, p, t, v), dim=-3)
-    a2 = torch.stack((t, v, v, q, p, p), dim=-3)
-    a3 = torch.stack((p, p, t, v, v, q), dim=-3)
-    a4 = torch.stack((a1, a2, a3), dim=-4)
+    # vpqt -> rgb mapping based on i
+    select = torch.tensor([[0, 2, 1, 1, 3, 0], [3, 0, 0, 2, 1, 1], [1, 1, 3, 0, 0, 2]], dtype=torch.long)
+    if img.device.type == "cuda":
+        select = select.pin_memory(img.device)
```

Comment:
I'm not sure about benefits of doing this, have you seen any significant perf improvements by doing this ?

Reply:
Not doing it will definitely cause a CPU/GPU sync, whether or not it will worsen your GPU utilization depends. A colleague and I went on a sync hunt in his augmentation pipeline and we got a measurable improvement after every last one of them was eliminated. If you can quickly refill the GPU queue (kernels longer than corresponding python code) then you don't need to worry about it.

Reply:
I'm running the perf benchmarks on the latest commit (https://github.com/pytorch/vision/pull/7754/commits/f3c2024a4c8741e3db8889aef9e76b78914ff83d) and removing `select.pin_memory` brings a speed-up. See `fn_v2` vs `fn_v2_plus`:
```
[------------- HSV -> RGB cpu torch.float32 -------------]
                         |  fn_v1  |  fn_v2  |  fn_v2_plus
1 threads: -----------------------------------------------
      (3, 400, 300)      |    7.7  |    3.2  |      3.2   
      (3, 640, 768)      |   45.6  |   14.7  |     14.8   
      (16, 3, 400, 300)  |  362.9  |   76.9  |     76.9   

Times are in milliseconds (ms).

[------------- HSV -> RGB cuda torch.float32 -------------]
                         |  fn_v1   |  fn_v2  |  fn_v2_plus
1 threads: ------------------------------------------------
      (3, 400, 300)      |   487.8  |  292.4  |    266.5   
      (3, 640, 768)      |   499.9  |  291.2  |    265.3   
      (16, 3, 400, 300)  |  2354.8  |  316.1  |    299.5   

Times are in microseconds (us).
```
[Source, revision 4](https://gist.github.com/vfdev-5/a6abf2d81a1c06e41392999fb674ce22#file-bench_hsv2rgb_pr_7753-py)

Can you share your findings such that I could reproduce your improvements with `select.pin_memory` code ?

Reply:
I was not able to reproduce the issue in a standalone example, so I have removed that bit of the code and instantiated the tensor on the device normally in the latest commit. If this issue is not solved already, it ought to be dealt with upstream in PyTorch anyway.